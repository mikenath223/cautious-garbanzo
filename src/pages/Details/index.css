.details-main {
  margin-block: 2.5rem;
}

.city-notes-input {
  padding-inline: 1rem;
  padding-block: 0.5rem;
  border: 1px solid var(--bg-orange);
  margin-right: 0.5rem;
}

.city-form-notes {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  align-items: flex-start;
  margin-top: 4rem;
  margin-bottom: 2rem;
}
