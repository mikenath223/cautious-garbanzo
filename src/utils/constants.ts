import { IWeatherInfo } from "./types"

export const initFavCities: string[] = [
  "Tokyo",
  "Delhi",
  "Shanghai",
  "Dhaka",
  "Sao Paulo",
  "Mexico City",
  "Cairo",
  "Beijing",
  "Mumbai",
  "Osaka",
  "Chongqing",
  "Karachi",
  "Kinshasa",
  "Lagos",
  "Istanbul"
]

export const defaultCity: IWeatherInfo = {
  "request": {
    "type": "City",
    "query": "New York, United States of America",
  },
  "location": {
    "name": "New York",
    "country": "United States of America",
    "region": "New York",
    "localtime": "2019-09-07 08:14",
  },
  "current": {
    "observation_time": "12:14 PM",
    "temperature": 13,
    "weather_code": 113,
    "weather_icons": [
      "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
    ],
    "weather_descriptions": [
      "Sunny"
    ],
    "wind_speed": 0,
    "wind_degree": 349,
    "wind_dir": "N",
    "pressure": 1010,
    "precip": 0,
    "humidity": 90,
    "cloudcover": 0,
    "feelslike": 13,
    "uv_index": 4,
    "visibility": 16
  }
}

export const defaultNote = {
  id: "01",
  text: "Welcome"
}
