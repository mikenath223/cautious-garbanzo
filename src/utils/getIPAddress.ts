const fetchIPAddress = async () => {
  try {
    const response = await fetch('https://ipapi.co/json')
    const data = await response.json()
    return data
  } catch (err) {
    console.error(err)
  }
}

export default fetchIPAddress
