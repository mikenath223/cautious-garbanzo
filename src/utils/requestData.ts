import { IWeatherInfo, PromiseFulfilledResult } from "./types"

export const getCityWeather = async (city: string, isCheckCache: boolean): Promise<IWeatherInfo> => {
  try {
    const currentDate = new Date()
    const localStorageKey = `${city.toLowerCase().trim()}-${currentDate.getHours()}-${currentDate.getDate()}-${currentDate.getMonth()}-${currentDate.getFullYear()}`
    const cachedCityData = localStorage.getItem(localStorageKey)
    if (cachedCityData && isCheckCache) {
      const weatherData = JSON.parse(cachedCityData)
      return weatherData
    }
    const url = `http://api.weatherstack.com/current?access_key=04c73e2c81cd37fd215de627927cdd44&query=${city}`;
    const response = await fetch(url)
    const result = await response.json()
    if (result.error) {
      return result?.error?.info
    }
    localStorage.setItem(localStorageKey, JSON.stringify(result))
    return result
  } catch (error) {
    const errMsg = `Failed to retrieve ${city.trim()} data!`
    throw new Error(errMsg)
  }
}

const sortCities = (cities: string[]) => cities.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))

export const getFavsCityWeather = async (cities: string[]) => {
  const fetchRequests = sortCities(cities).map(city => getCityWeather(city, true));

  try {
    const citiesInfo = await Promise.allSettled(fetchRequests)
    const data = citiesInfo.filter(data => data.status === "fulfilled").map(data => (data as PromiseFulfilledResult<IWeatherInfo>).value)
    return data
  } catch (_) {
    return []
  }
}