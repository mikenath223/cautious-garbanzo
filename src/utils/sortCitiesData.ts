import { IWeatherInfo } from "./types";

const sortCitiesData = (cities: IWeatherInfo[] | undefined) => cities?.sort((a, b) => a.location.name.toLowerCase().localeCompare(b.location.name.toLowerCase()))

export default sortCitiesData
