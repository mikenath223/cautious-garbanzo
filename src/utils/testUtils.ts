import { render } from '@testing-library/react';

export const withRender = (component: JSX.Element) =>
  render(component);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => { };
  localStorage.getItem = (key: string) => null;
}