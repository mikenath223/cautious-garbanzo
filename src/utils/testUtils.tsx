import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export const withRender = (component: JSX.Element) =>
  render(<MemoryRouter>{component}</MemoryRouter>);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => {};
  localStorage.getItem = (key: string) => null;
};
