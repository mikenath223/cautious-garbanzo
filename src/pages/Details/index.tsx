import { Link } from "react-router-dom";
import useLocalStorage from "../../hooks/useLocalStorage";
import { INote, IWeatherInfo } from "../../utils/types";
import { useMemo, useState } from "react";
import WeatherDetails from "../../components/WeatherDetails";
import "./index.css";
import Button from "../../components/Button";
import Notes from "../../components/Notes";
import { v4 as uuidv4 } from "uuid";

export default function Details() {
  const selectedCity = localStorage.getItem("selected-city");
  const city: IWeatherInfo | undefined = selectedCity
    ? JSON.parse(selectedCity)
    : undefined;
  const notesKey = `${city?.location?.name}-notes`;

  const [selectedCityNotes, setSelectedCityNotes] = useLocalStorage<
    INote[] | undefined
  >(notesKey, undefined);
  const [favCitiesTemp, setFavCitiesTemp] = useLocalStorage<
    IWeatherInfo[] | undefined
  >("favorite-cities", undefined);
  const [inputNote, setInputNote] = useState("");
  const [editNote, setEditNote] = useState<INote | null>(null);

  const isCityFavorites = useMemo(() => {
    return favCitiesTemp
      ? favCitiesTemp.some(
          (favCity) => favCity.location.name === city?.location?.name
        )
      : false;
  }, [city?.location?.name, favCitiesTemp]);

  const onSetNote = (event: React.FormEvent<HTMLTextAreaElement>) => {
    setInputNote(event.currentTarget.value);
  };

  const resetNote = () => {
    setEditNote(null);
    setInputNote("");
  };

  const onSaveNote = (e: React.SyntheticEvent) => {
    e.preventDefault();
    if (editNote) {
      const newNotes = selectedCityNotes?.map((note) => {
        if (note.id === editNote.id) {
          return {
            ...note,
            text: inputNote,
          };
        }
        return note;
      });
      setSelectedCityNotes(newNotes);
      resetNote();
      return;
    }
    const newNote = {
      id: uuidv4(),
      text: inputNote,
    };
    const newNotes = [newNote, ...(selectedCityNotes || [])];
    setSelectedCityNotes(newNotes);
    setInputNote("");
  };

  const onEditNote = (note: INote) => {
    setEditNote(note);
    setInputNote(note.text);
  };

  const onDeleteNote = (note: INote) => {
    resetNote();
    const newNotes = selectedCityNotes?.filter(
      (cityNote) => cityNote.id !== note.id
    );
    setSelectedCityNotes(newNotes);
  };

  const onAddFavorites = () => {
    if (city && favCitiesTemp) {
      const newFavs = [city, ...(favCitiesTemp || [])];
      setFavCitiesTemp(newFavs);
    }
  };

  const onRemoveFavorites = () => {
    const newFavs = favCitiesTemp?.filter(
      (favCity) => favCity.location.name !== city?.location.name
    );
    setFavCitiesTemp(newFavs);
  };

  return (
    <main className="details-main max-width-inhibitor">
      {city?.location ? (
        <>
          <WeatherDetails
            city={city}
            isFavorites={isCityFavorites}
            onAddFavorites={onAddFavorites}
            onRemoveFavorites={onRemoveFavorites}
          />
          <form onSubmit={onSaveNote} className="city-form-notes">
            <textarea
              placeholder="Save notes on city..."
              onChange={onSetNote}
              value={inputNote}
              className="city-notes-input"
              required
            ></textarea>
            <Button type="submit">Submit</Button>
          </form>
          <ul className="city-notes">
            {selectedCityNotes?.map((note) => (
              <Notes
                key={note.id}
                note={note}
                onEditNote={onEditNote}
                onDeleteNote={onDeleteNote}
              />
            ))}
          </ul>
        </>
      ) : (
        <h1>
          There was an error. Please go to <Link to="/">home</Link>
        </h1>
      )}
    </main>
  );
}
