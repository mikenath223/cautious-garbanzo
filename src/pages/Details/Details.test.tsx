import { withRender } from "../../utils/testUtils";
import { MemoryRouter } from "react-router-dom";
import Details from "./index";

describe("Searchbar", function () {
  it("should match snapshot", () => {
    const { container } = withRender(
      <MemoryRouter>
        <Details />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
