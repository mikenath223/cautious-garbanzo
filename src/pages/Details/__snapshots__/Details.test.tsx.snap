// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar should match snapshot 1`] = `
<div>
  <main
    class="details-main max-width-inhibitor"
  >
    <h1>
      There was an error. Please go to 
      <a
        href="/"
      >
        home
      </a>
    </h1>
  </main>
</div>
`;
